manager/src/main/java/com/redhat/service/bridge/manager/workers/resources/AbstractWorker.java

```diff
@@ -43,6 +43,7 @@ public T handleWork(Work work) {
         T managedResource = load(work);
         if (Objects.isNull(managedResource)) {
             //Work has been scheduled but cannot be found. Something (horribly) wrong has happened.
+            workManager.complete(work);
```

Comment:
@manstis We need to remove the `@Transactional` annotation from `BridgeServiceImpl.createBridge()` as it's currently leading to dirty reads in our `AbstractWorker`. More on dirty reads here: https://www.sqlshack.com/dirty-reads-and-the-read-uncommitted-isolation-level/

The problem is that the `ManagedResource` instance and `Work` are being created in the same transaction. So the following is possible:

- Transaction 1: writes the `Work` item to the table (but not yet the `ManagedResource`)
- Transaction 2 (work queue), reads the `Work` item and starts to execute. It checks for the `ManagedResource` which has not yet been written and sees it as missing. It then deletes the `Work` item
- Transaction 1: commits 

To fix this, I'd suggest the following changes:

Update to look `BridgeServiceImpl.createBridge` to look like this:

```
    @Override
    public Bridge createBridge(String customerId, BridgeRequest bridgeRequest) {
        
        Bridge bridge = doCreateBridge(customerId, bridgeRequest);

        LOGGER.info("Bridge with id '{}' has been created for customer '{}'", bridge.getId(), bridge.getCustomerId());

        workManager.schedule(bridge);

        return bridge;
    }
```

`BridgeServiceImpl.doCreateBridge()` to look like this:

```
@Transactional
    protected Bridge doCreateBridge(String customerId, BridgeRequest bridgeRequest) {
if (bridgeDAO.findByNameAndCustomerId(bridgeRequest.getName(), customerId) != null) {
            throw new AlreadyExistingItemException(String.format("Bridge with name '%s' already exists for customer with id '%s'", bridgeRequest.getName(), customerId));
        }
        
        Bridge bridge = bridgeRequest.toEntity();
        bridge.setStatus(ManagedResourceStatus.ACCEPTED);
        bridge.setSubmittedAt(ZonedDateTime.now(ZoneOffset.UTC));
        bridge.setCustomerId(customerId);
        bridge.setShardId(shardService.getAssignedShardId(bridge.getId()));
        bridgeDAO.persist(bridge);

        return bridge;
}
```
This will mean that the `ManagedResource` instance is comitted to the DB (or not) before the `Work` item.

Reply:
Perfect sense and consistent with deletion of Bridges and the creation/deletion of Processors.

Reply:
Fixed in new commit.

Reply:
@manstis I wanted to come back to update you on this after a quick discussion between @lucamolteni and I on Friday.

@lucamolteni correctly pointed out that Postgres doesn't support `READ_UNCOMMITTED` transaction isolation, so the dirty ready problem I called out is not correct. More here: https://www.postgresql.org/docs/12/transaction-iso.html

Instead, the problem was being caused by the immediate invocation of the vertX worker, which you have addressed in #480 

When the vertX worker was called immediately, because it was not within the same transaction as the one writing the initial data to the DB for the `Bridge` & ` Work`, it did not find the resource and hence reported that it did not exist. This is what I mistook for a dirty read.

Looking at the code we have now, I see no reason why we can't simply collapse the `doCreateBridge` into `createBridge` and `doDeleteBridge` into `deleteBridge`, placing the `Transactional` annotation back onto the `createBridge`/`deleteBridge` methods.

Apologies for making the wrong suggestion in the first place.

cheers,

Rob

Reply:
LOL.. thank-you for recollecting better than myself. 

I had an impossible time trying to explain it to @lucamolteni earlier.

The change herein makes provisioning of bridges consistent with that of processors and with the de-provisioning of both. We could probably now collapse the methods (as we no longer execute the work immediately). However, since this PR brings consistency, I'd prefer to do as a separate task.